pkg/socketenricher/bpf/socket-enricher.bpf.c

```diff
@@ -31,29 +32,65 @@ struct {
 
 const volatile bool disable_bpf_iterators = 0;
 
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 1024);
+	__type(key, __u64); // pid_tgid
+	__type(value, struct sockets_value);
+} ig_tmp_sockets_value SEC(".maps");
+
+static const struct sockets_value empty_sockets_value = {};
+
 static __always_inline void insert_current_socket(struct sock *sock)
 {
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
 	struct sockets_key socket_key = {
 		0,
 	};
 	prepare_socket_key(&socket_key, sock);
 
-	struct sockets_value socket_value = {
-		0,
-	};
+	if (bpf_map_update_elem(&ig_tmp_sockets_value, &pid_tgid,
+				&empty_sockets_value, BPF_NOEXIST))
+		return;
+
+	struct sockets_value *socket_value =
+		bpf_map_lookup_elem(&ig_tmp_sockets_value, &pid_tgid);
+	if (!socket_value)
+		return;
+
 	// use 'current' task
 	struct task_struct *task = (struct task_struct *)bpf_get_current_task();
-	socket_value.mntns = (u64)BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
-	socket_value.pid_tgid = bpf_get_current_pid_tgid();
-	socket_value.uid_gid = bpf_get_current_uid_gid();
-	bpf_get_current_comm(&socket_value.task, sizeof(socket_value.task));
-	socket_value.sock = (__u64)sock;
+	struct task_struct *parent = BPF_CORE_READ(task, real_parent);
+	struct fs_struct *fs = BPF_CORE_READ(task, fs);
+	struct file *exe_file = BPF_CORE_READ(task, mm, exe_file);
+	socket_value->mntns =
+		(u64)BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
+	socket_value->pid_tgid = bpf_get_current_pid_tgid();
+	socket_value->uid_gid = bpf_get_current_uid_gid();
+	bpf_get_current_comm(&socket_value->task, sizeof(socket_value->task));
+	if (parent != NULL) {
+		bpf_probe_read_kernel(&socket_value->ptask,
+				      sizeof(socket_value->ptask),
+				      parent->comm);
+		socket_value->ppid = (__u32)BPF_CORE_READ(parent, tgid);
+	}
+	char *cwd = get_path_str(&fs->pwd);
```

Comment:
The implementation of `get_path_str` writes on its internal buffer from the end, since it does not know in advance the size of the string. So I think the double copy cannot be avoided with this algorithm.

But yes, we could avoid the gadget developer having to do it explicitly.